storage/ndb/include/ndbapi/NdbInterpretedCode.hpp

```diff
@@ -195,6 +275,88 @@ class NdbInterpretedCode
+  int or_reg(Uint32 RegDest, Uint32 RegSource1, Uint32 RegSource2);
+  int xor_reg(Uint32 RegDest, Uint32 RegSource1, Uint32 RegSource2);
+  int mod_reg(Uint32 RegDest, Uint32 RegSource1, Uint32 RegSource2);
+  int not_reg(Uint32 RegDest, Uint32 RegSource1);
+
+  /* Register arithmetic
+   * -------------------
+   * These instructions provide arithmetic operations on the
+   * interpreter's registers. They have the same use as the
+   * one with two source registers except that they use a
+   * constant as the second part of the operation.
+   *
+   * *RegDest= *RegSouce1 <operator> *Constant
+   *
+   * Space required   Buffer    Request message
+   *   add_const_reg  1 word    1 word
+   *   sub_const_reg  1 word    1 word
+   *   lshift_const_reg 1 word  1 word
+   *   rshift_const_reg 1 word  1 word
+   *   mul_const_reg  1 word    1 word
+   *   div_const_reg  1 word    1 word
+   *   and_const_reg  1 word    1 word
+   *   or_const_reg   1 word    1 word
+   *   xor_const_reg  1 word    1 word
+   *   mod_const_reg  1 word    1 word
+   *
+   * @param RegDest Register to store operation result in
+   * @param RegSource1 Register to use as LHS of operator
+   * @param Constant to use as RHS of operator
+   * @return 0 if successful, -1 otherwise
+   */
+  int add_const_reg(Uint32 RegDest, Uint32 RegSource1, Uint16 Constant);
+  int sub_const_reg(Uint32 RegDest, Uint32 RegSource1, Uint16 Constant);
+  int lshift_const_reg(Uint32 RegDest, Uint32 RegSource1, Uint16 Constant);
+  int rshift_const_reg(Uint32 RegDest, Uint32 RegSource1, Uint16 Constant);
+  int mul_const_reg(Uint32 RegDest, Uint32 RegSource1, Uint16 Constant);
+  int div_const_reg(Uint32 RegDest, Uint32 RegSource1, Uint16 Constant);
+  int and_const_reg(Uint32 RegDest, Uint32 RegSource1, Uint16 Constant);
+  int or_const_reg(Uint32 RegDest, Uint32 RegSource1, Uint16 Constant);
+  int xor_const_reg(Uint32 RegDest, Uint32 RegSource1, Uint16 Constant);
+  int mod_const_reg(Uint32 RegDest, Uint32 RegSource1, Uint16 Constant);
+
+  /* Move from heap memory to register
+   * ---------------------------------
+   * These instructions provide possibilities to read the memory
+   * inserted from reads of columns and load it into registers.
+   * One can read 1 byte treated as an Uint8, 2 bytes treated as
+   * Uint16, 4 bytes treated as Uint32 and 8 bytes treated as
+   * Int64.
+   *
+   * *RegDest= *RegSouce1 <operator> *Constant
+   *
+   * Space required   Buffer    Request message
+   *   read_u8_to_reg 1 word    1 word
+   *   read_u16_to_reg 1 word   1 word
+   *   read_u32_to_reg 1 word   1 word
+   *   read_int64_to_reg 1 word 1 word
+   *
+   * @param RegDest Register to store operation result in
+   * @param RegSource1 Register to use as LHS of operator
+   * @param Constant to use as RHS of operator
+   * @return 0 if successful, -1 otherwise
+   */
+  int read_u8_to_reg(Uint32 RegDest, Uint16 memory_offset);
```

Comment:
These four commands allow to copy from memory based on an immediate value or, from the perspective of the interpreted code, constant.

For me the offsets are result of a interpreted code computation, and thus they need to be register values not constants.

Reply:
Ok, will ensure that these variants are also supported, haven't decided on how yet, but will fix it

Reply:
Very nice. thank you, Michael! If space is scarce you could also *replace* the existing commands with ones that use a register instead of a constant, instead of *adding* four more commands. I would expect that read_u*_to_reg with a constant offset is maybe less often? And if necessary, one could do it via a register.

Reply:
You asked me to repeat requests in public comments, so I repeat it here. Just as per request not to annoy you. If you decide against it, that is okay, of course, and I use the replacements:

Additional branch_ne/eq_const and branch_ne/eq_zero would be helpful as well. If only one then branch_ne_const, which I need most often and could replace branch_ne_zero already.

Reason: While it is possible to load constants into registers first and then use branch_ne these are done often in our algorithms, and these are especially necessary in code paths that hit often and that can usually be decided very quickly. We have some special cases just to handle the most frequent cases very quickly without having to resort to the binary search (which handles the general case). For those test I have to compare a length value to the constants 0, 1, and 2 in succession. Here branch_ne_const would be nice, as it wouldn't need a separate command and register.

In general, I could also do the necessary code changes now and provide a commit to introduce those. Not next week, but the week after. At the same time I figure, the addition is relatively straightforward, while the decision which to add, needs more consideration.

Thanks for your time, Max

Reply:
Yep, I will add new Branch on zero/nonzero instructions as well and with the new opcodes I don't have a problem finding opcodes to use. So it won't take much time to add a few more branch instructions as well

Reply:
I have added checks that registers are not out of bound, also done some internal changes to ensure
it is clear which opcodes are still free